gadopt/nullspaces.py

```diff
@@ -0,0 +1,216 @@
+    $$
+
+    followed by integration by parts:
+
+    $$
+        int_Gamma -bb n * q nabla p ds + int_Omega nabla q cdot nabla p dx +
+        int_Gamma bb n * hatk q g "Di" rho chi c_p/(c_v gamma) p dx -
+        int_Omega nabla q * hatk g "Di" rho chi c_p/(c_v gamma) p dx = 0
+    $$
+
+    This elliptic equation can be solved with natural boundary conditions by imposing our
+    original equation above, which eliminates all boundary terms:
+
+    $$
+      int_Omega nabla q * nabla p dx - int_Omega nabla q * hatk g "Di" rho chi c_p/(c_v gamma) p dx = 0.
+    $$
+
+    However, if we do so on all boundaries we end up with a system that has the same
+    null space, as the one we are after (note that we ended up merely testing the
+    original equation with $nabla q$). Instead we use the fact that the gradient of
+    the null mode is always vertical, and thus the null mode is constant at any
+    horizontal level (geoid), specifically the top surface. Choosing any nonzero
+    constant for this surface fixes the arbitrary scalar multiplier of the null
+    mode. We choose the value of one and apply it as a Dirichlet boundary condition.
+
+    Note that this procedure does not necessarily compute the exact null space of the
+    *discretised* Stokes system. In particular, since not every test function
+    $v in V$, the velocity test space, can be written as $v=nabla q$ with $q in W$,
+    the pressure test space, the two terms do not necessarily exactly cancel when
+    tested with $v$ instead of $nabla q$ as in our final equation. However, in
+    practice the discrete error appears to be small enough, and providing this
+    null space gives an improved convergence of the iterative Stokes solver.
+    """
+    W = fd.FunctionSpace(mesh=W.mesh(), family=W.ufl_element())
+    q = fd.TestFunction(W)
+    p = fd.Function(W, name="pressure_nullspace")
+
+    # Fix the solution at the top boundary
+    bc = fd.DirichletBC(W, 1.0, top_subdomain_id)
+
+    F = fd.inner(fd.grad(q), fd.grad(p)) * fd.dx
+
+    k = upward_normal(W.mesh())
+
+    F += (
+        -fd.inner(fd.grad(q), k * approximation.dbuoyancydp(p, fd.Constant(1.0)) * p)
+        * fd.dx
+    )
+
+    fd.solve(F == 0, p, bcs=bc)
+    return p
+
+
+def create_stokes_nullspace(
+    Z: fd.functionspaceimpl.WithGeometry,
+    closed: bool = True,
+    rotational: bool = False,
+    translations: list[int] | None = None,
+    ala_approximation: AnelasticLiquidApproximation | None = None,
+    top_subdomain_id: str | int | None = None,
+) -> fd.nullspace.MixedVectorSpaceBasis:
+    """Create a null space for the mixed Stokes system.
+
+    Arguments:
+      Z: Firedrake mixed function space associated with the Stokes system
+      closed: Whether to include a constant pressure null space
+      rotational: Whether to include all rotational modes
+      translations: List of translations to include
```

Comment:
While we're at it, probably good to document what the `translations` list should look like?

Reply:
thanks!